google/cloud/spanner_dbapi/connection.py

```diff
@@ -178,6 +182,42 @@ def run_prior_DDL_statements(self):
 
         return self.__handle_update_ddl(ddl_statements)
 
+    def run_statement(self, sql, params, param_types):
+        """Run single SQL statement in begun transaction.
+
+        This method is never used in autocommit mode. In
+        !autocommit mode however it remembers every executed
+        SQL statement with its parameters.
+
+        :type sql: :class:`str`
+        :param sql: SQL statement to execute.
+
+        :type params: :class:`dict`
+        :param params: Params to be used by the given statement.
+
+        :type param_types: :class:`dict`
+        :param param_types: Statement parameters types description.
+
+        :rtype: :class:`google.cloud.spanner_v1.streamed.StreamedResultSet`,
+                :class:`google.cloud.spanner_dbapi.checksum.ResultsChecksum`
+        :returns: Streamed result set of the statement and a
+                  checksum of this statement results.
+        """
+        transaction = self.transaction_checkout()
+
+        statement = {
+            "sql": sql,
+            "params": params,
+            "param_types": param_types,
+            "checksum": ResultsChecksum(),
+        }
+        self._statements.append(statement)
+
+        return (
+            transaction.execute_sql(sql, params, param_types=param_types),
+            statement["checksum"],
+        )
```

Comment:
As we need to track statements executed in the transaction, and several cursors can run statements in the same transaction from different points, I'm proposing to add a method into `Connection` class. It will be saving all the executed statements in the connection attribute.

In the same time, results are fetched on cursors, so it makes sense to pass a results streaming iterator as well as a pointer to a checksum back into cursor. So when a cursor is fetching the next row, it also updates a results checksum of the statement.